x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/snapshot/upgrader/SnapshotUpgradePredicate.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ml.job.snapshot.upgrader;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.persistent.PersistentTasksCustomMetadata;
+import org.elasticsearch.xpack.core.ml.job.snapshot.upgrade.SnapshotUpgradeState;
+import org.elasticsearch.xpack.core.ml.job.snapshot.upgrade.SnapshotUpgradeTaskState;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.xpack.ml.job.task.OpenJobPersistentTasksExecutor.checkAssignmentState;
+
+public class SnapshotUpgradePredicate implements Predicate<PersistentTasksCustomMetadata.PersistentTask<?>> {
+    private final boolean waitForCompletion;
+    private final Logger logger;
+    private volatile Exception exception;
+    private volatile String node = "";
+    private volatile boolean shouldCancel;
+    private volatile boolean isCompleted;
+
+    public SnapshotUpgradePredicate(boolean waitForCompletion, Logger logger) {
+        this.waitForCompletion = waitForCompletion;
+        this.logger = logger;
+    }
+
+    public Exception getException() {
+        return exception;
+    }
+
+    public String getNode() {
+        return node;
+    }
+
+    public boolean isShouldCancel() {
+        return shouldCancel;
+    }
+
+    public boolean isCompleted() {
+        return isCompleted;
+    }
+
+    @Override
+    public boolean test(PersistentTasksCustomMetadata.PersistentTask<?> persistentTask) {
+        // Persistent task being null means it has been removed from state, and is now complete
+        if (persistentTask == null) {
+            isCompleted = true;
+            return true;
+        }
```

Comment:
In all my testing, the task is only `null` when it has been removed from cluster state. Since this predicate runs AFTER we have confirmed the task has been added to state (the `start task API`), it is good to assume that `null` is removal and thus is completion.